### PR TITLE
MGMT-17673: [Docs] document image building

### PR DESCRIPTION
**Pull request by fzdarsky**

### CODE REVIEW

File: docs/user/building-images.md [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+## Building and Publishing OS Images and Disk Images
+
+This section describes the generic process for building an OS image (bootc) that contains the Flight Control agent and building an OS disk image for flashing to physical devices. Later sections then describe considerations when building for specific virtualization and bare metal provisioning environments.
+
+Before you start, ensure you have installed the following prerequisites:
+
+* `flightctl` CLI latest version ([installation guide](gettting-started.md#-installing-the-flight-control-cli))
+* `podman` version 4.9 or higher ([installation guide](https://podman.io/docs/installation))
```
Comment (fzdarsky) [inline docs/user/building-images.md:37]: Must be version 5.0 or higher for OCI artifacts support.

File: docs/user/building-images.md [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+```console
+sudo podman manifest create ${OCI_IMAGE_REPO}-unified:${OCI_IMAGE_TAG}
+sudo podman manifest add \
+    ${OCI_IMAGE_REPO}-unified:${OCI_IMAGE_TAG} docker://${OCI_IMAGE_REPO}:${OCI_IMAGE_TAG}
+sudo podman manifest add \
+    --artifact --artifact-type application/x-raw-disk-image \
+    --os=linux --arch=x86_64 --annotation disktype=raw \
+    ${OCI_IMAGE_REPO}-unified:${OCI_IMAGE_TAG} $(pwd)/output/image/disk.raw
+sudo podman manifest push --all \
```
Comment (fzdarsky) [inline docs/user/building-images.md:185]: Need to add the image signing `--sign-by-sigstore-private-key ./signingkey.private`.

File: docs/user/building-images.md [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+sudo podman run --rm -it --privileged --pull=newer --security-opt label=type:unconfined_t \
+    -v $(pwd)/output:/output -v /var/lib/containers/storage:/var/lib/containers/storage \
+    quay.io/centos-bootc/bootc-image-builder:latest \
+    --type vmdk ${OCI_IMAGE_REPO}:${OCI_IMAGE_TAG}
+```
+
+Once `bootc-image-builder` completes, you can find the disk image under `$(pwd)/output/vmdk/disk.vmdk`.
+
+## Best Practices when Building Images
```
Comment (hexfusion) [inline docs/user/building-images.md:240]:
```suggestion
## Best Practices When Building Images
```

File: docs/user/building-images.md [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+RUN dnf -y copr enable @redhat-et/flightctl && \
+    dnf -y install flightctl-agent && \
+    dnf -y clean all && \
+    systemctl enable flightctl-agent.service
+
+ADD config.yaml /etc/flightctl/
+```
+
+[!NOTE]
+If you have used Podman or Docker before to build application containers, you will notice this is a regular `Containerfile`, with the only difference that the base image referenced in `FROM` is bootable container (bootc) image. That means it already contains a Linux kernel. This allows you to reuse your existing, standard container build tools and workflows.
```
Comment (hexfusion) [inline docs/user/building-images.md:104]:
```suggestion
If you have used Podman or Docker before to build application containers, you will notice this is a regular `Containerfile`, with the only difference that the base image referenced in `FROM` is bootable container (bootc) image. That means it already contains a Linux kernel. This allows you to reuse existing standard container build tools and workflows.
```

File: docs/user/building-images.md [hunk 5]
```diff
@@ -0,0 +1,268 @@
+# Building OS Images
+
+## Understanding OS Images and the Image Build Process
+
+Image-based OSes allow the whole OS to be versioned, deployed, and updated as a single unit. This reduces operational risk:
+
+* It minimizes potential drift between what has been thoroughly tested and what is deployed to a large number of devices.
+* It minimizes the risk of failed updates that require expensive maintenance or replacement though transactional updates and rollbacks.
```
Comment (hexfusion) [inline docs/user/building-images.md:8]:
```suggestion
* It minimizes the risk of failed updates that require expensive maintenance or replacement through transactional updates and rollbacks.
```

File: docs/user/building-images.md [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+
+Prefer adding configuration to the OS image itself at build-time, so they get tested, distributed, and updated together ("lifecycle-bound"). There are cases when this is not feasible or desirable, in which case use Flight Control's approach to dynamically configure devices at runtime instead:
+
+* configuration that is deployment- or site-specific (e.g. a hostname or a site-specific network credential)
+* secrets that are not secure to distribute via the image
+* application workloads that need to be added, updated, or deleted without reboot or on a faster cadence than the OS
+
+### Prefer Configuration in /usr over /etc
+
+Prefer placing  configuration files under `/usr` if the configuration is static and the application or service supports it. This way, configuration remains read-only and fully defined by the image (avoiding ostree's 3-way merge of `/etc` as potential source of drift). There are cases when this is not feasible or desirable:
```
Comment (hexfusion) [inline docs/user/building-images.md:252]:
```suggestion
Prefer placing  configuration files under `/usr` if the configuration is static and the application or service supports it. This way, configuration remains read-only and fully defined by the image (avoiding ostree's 3-way merge of `/etc` as a potential source of drift). There are cases when this is not feasible or desirable:
```

File: docs/user/building-images.md [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+1. Use late binding of the enrollment endpoint and enrollment certificates, injecting the enrollment certificate or even the whole agent configuration through `cloud-init` when provisioning the virtual device.
+2. Optionally, add the `open-vm-tools` guest tools to the image.
+3. Build a disk image of type "vmdk" instead of "raw".
+
+Create a file named `Containerfile` with the following content to build an OS image based on CentOS Stream 9 that includes the Flight Control agent and VMware guest tools, but no agent configuration:
+
+```console
+FROM quay.io/centos-bootc/centos-bootc:stream9
+
+RUN dnf -y copr enable @redhat-et/flightctl-dev centos-stream-9-x86_64 && \
```
Comment (celdrake) [inline docs/user/building-images.md:216]: In the previous section "### Building the OS Image (bootc)" the copr repo mentioned was `@redhat-et/flightctl`. The command is also different in that one specifies the centos stream including the architecture, and the other doesn't.


Should these be the same command, and generally they should target the non-dev repo?

Comment (fzdarsky) [reply docs/user/building-images.md:216]: Good catch, thank you! Yes, it should be the non-dev repo. It should also be without the chroot (centos-stream-9-x86_64), but that's something we have to fix still. Will keep it in until the fix lands.

**Review by avishayt — APPROVED** [review]

**Pull request merged**